tutorial: add second part of tutorial series for building DReddit

PascalPrecht opened the pull request:
This is now only part 2 of the tutorial

jrainville commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-1.md:
```diff
@@ -0,0 +1,192 @@
+
+<pre><code>$ embark run</code></pre>
+
+If there are any issues in the "Available Services" section of the dashboard, go back to our installation guide and make sure all tools are available on your machine.
```
Please add a link to the "installation guide"

PascalPrecht replied:
👍

jrainville commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-1.md:
```diff
@@ -0,0 +1,192 @@
+
+### Creating posts
+
+It's time to add our first method which will enable users to add new posts to the platform. For that we'll create method `createPost(bytes _description)` where `_description` are the bytes that represent the posts text.
```
```suggestion
It's time to add our first method which will enable users to add new posts to the platform. For that, we'll create the method `createPost(bytes _description)`, where `_description` are the bytes that represent the posts text.
```

PascalPrecht replied:
Wonderful! Fixed!

jrainville commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-1.md:
```diff
@@ -0,0 +1,192 @@
+  });
+}</code></pre>
+
+The first thing we do is creating an id for the post to be stored. We then use our `Post` struct to create a new post instance. Notice that we leverage the `postId` when storing the Post in our `posts` array. To set the owner we can take advantage of Solidity's global `msg` object which is available in every transaction. 
```
```suggestion
The first thing we do is creating an id for the post to be stored. We then use our `Post` struct to create a new post instance. Notice that we leverage the `postId` when storing the Post in our `posts` array. To set the owner, we take advantage of Solidity's global `msg` object which is available in every transaction. 
```

PascalPrecht replied:
👍

jrainville commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-1.md:
```diff
@@ -0,0 +1,192 @@
+
+### Up and down voting posts
+
+As mentioned earlier, Reddit allows for up and down voting topic posts. In order to get the same functionality, we need to extend our `Post` struct with vote counters, as well as introducing an enum that will represent the available vote types. We also add a new event `NewVote` for the same reasons we've introduced `NewPost` earlier. Once that is done, we can add a method that perform actual vote ...
```
```suggestion
As mentioned earlier, Reddit allows for up and down voting topic posts. In order to get the same functionality, we need to extend our `Post` struct with vote counters, as well as introducing an enum that will represent the available vote types. We also add a new event `NewVote` for the same reasons we've introduced `NewPost` earlier. Once that is done, we can add a method that performs actual votes.
```

PascalPrecht replied:
Good catch!

jrainville commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-1.md:
```diff
@@ -0,0 +1,192 @@
+  return uint8(post.voters[msg.sender]);
+}</code></pre>
+
+And with that, our Smart Contract is pretty much done! Just to make sure that everything is compiling smoothly, we can execute `embark run` in case there's no existing instance watching our work already.
```
Since this is only smart contract, you can recommend using `embark build --contract`

alaibe commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-1.md:
```diff
@@ -0,0 +1,192 @@
+
+## Creating the Smart Contract
+
+Alright, next up we want to create the brain of our application, which is a Smart Contract that enables creating posts and votes. We're going to build it up step by step and afterwards we'll add some tests to ensure our code is actually working.
```
Would it be good to link to solidity documentation?

PascalPrecht replied:
added!

alaibe commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md:
```diff
@@ -0,0 +1,194 @@
+
+### Requiring Smart Contract instances
+
+When running tests, Embark adds a couple of custom functions and objects to the global scope, which are useful or even necessary. One of those functions is a custom `require()` that lets us import Smart Contract instances from an Embark specific path.
```
which are ~~useful or even~~ necessary
We could probably explain why it is necessary or link to the doc that explains it

PascalPrecht replied:
Hm.. yes, but it's actually only necessary because we're providing those custom paths to import from. But theoretically, we could have `embark test` output compiled artifacts on the disc as well and have users import from there. So the necessity kind of comes with the fact that we introduce this thing in the first place.

Is there any other reason we introduce a custom `require()`?

alaibe commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md:
```diff
@@ -0,0 +1,194 @@
+
+## Testing correctness of data
+
+Another good test case would be to check if the stored data such as the description bytes, the owner etc.  resolve back to the correct data. Notice that this is slightly different from what we're testing in our previous test - there we're testing the description bytes emitted by the `NewPost` event. To test this we take advantage of the `postId` created in the previous test, which is available gl ...
```
Not sure, it feels weird
```the owner etc.  resolve ```
=>
```the owner etc...  resolve ```

jrainville replied:
The way Pascal wrote it is correct as far as I know

alaibe commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md:
```diff
@@ -0,0 +1,194 @@
+  it("should be able to vote in a post", async () => {
+    const receipt = await DReddit.methods.vote(postId, 1).send();
+    const Vote = receipt.events.NewVote;
+    assert.equal(Vote.returnValues.owner, accounts[0]);
```
Shall we explain that accounts[0] is the default account used.
Or make it explicit in the `send`?

PascalPrecht replied:
Yes I'll add a comment about that. Good point!

PascalPrecht commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-1.md:
```diff
@@ -0,0 +1,192 @@
+
+If you've read our guide on [Creating Applications](/docs/create_project.html) or our last tutorial on [Building Smart Contract only apps](/news/2019/01/22/building-smart-contract-only-dapps/), you know that Embark comes with a `new` command to scaffold an application. We're going to do exactly that, but first we need to make sure Embark is installed. For a complete guide on installing Embark, he ...
+
+<pre><code>$ npm install -g embark@next</code></pre>
```
👍

PascalPrecht commented:
@alaibe @jrainville addressed your comments in #136 

#136 contains only the first part of the series so we can merge and publish it today. We're leaving this open for the second (and prolly first part), as I don't wanna lose your comments.

jrainville commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md:
```diff
@@ -0,0 +1,194 @@
+For example, in order to get an instance of our `DReddit` Smart Contract within the test, we add the following line to our spec file:
+
+
+<pre class="javascript"><code>const DReddit = embark.require('Embark/contracts/DReddit');</code></pre>
```
The user can just do `require` without the `embark.`.  Might be more confusing, but you can explain it

PascalPrecht replied:
I actually explain it a moment earlier, so I'm gonna remove `embark` from there! 👌

jrainville commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md:
```diff
@@ -0,0 +1,194 @@
+contract('DReddit', () => {
+  ...
+  it ('should be able to create a post and receive it via contract event', async () => {
+    let receipt = await DReddit.methods.createPost(web3.utils.fromAscii(ipfsHash)).send();
```
```suggestion
    const receipt = await DReddit.methods.createPost(web3.utils.fromAscii(ipfsHash)).send();
```

PascalPrecht replied:
👍

jrainville commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md:
```diff
@@ -0,0 +1,194 @@
+
+Alright, next up let's quickly test if our `canVote()` method works the way as expected. As voting on posts that don't exist should never work, we will simply call `canVote()` on a post id that doesn't exist. This test is pretty straight forward:
+
+<pre class="javascript"><code>contract('DReddit', () => {
```
IMO, you don't have to put the `contract(...` everywhere. After one or to example, users probably got the gist of it.

PascalPrecht replied:
Yea, okay will roll back on that one.

jrainville commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md:
```diff
@@ -0,0 +1,194 @@
+
+We also want to make sure that `canVote()` resolves to `true` in case a user can indeed vote a certain post. We can again reuse the `postId` that we've stored earlier:
+
+<pre class="javascript"><code>it('should be able to vote in a post if account hasn\'t voted before', async () => {
```
```suggestion
<pre class="javascript"><code>it('should be able to vote in a post if account has not voted before', async () => {
```
Very aesthetic change. Backslashes are ugly

PascalPrecht replied:
Ha okay

jrainville commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md:
```diff
@@ -0,0 +1,194 @@
+
+ > All tests passed</code></pre>
+ 
+ Awesome! If you run into any issues, check out the repository with all steps recorded [here](TODO). In the next and last part of this series, we'll be building a front-end for our DReddit app using React. Until then, feel free to add more tests as you like!
```
Nice. Was just gonna recommend putting a link to the repo where all the tests are implemented.

PascalPrecht replied:
Done!

PascalPrecht commented on source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md:
```diff
@@ -0,0 +1,194 @@
+
+});</code></pre>
+
+Inside this code block we'll be putting dedicated test cases. The `contract()` function can be considered a "grouping" functionality to group tests, if you will. It comes with a couple of characteristics specific to Embark, but generally one can say that `contract()` is not only the equivalent to, but an extension of Mocha's [describe()]() function.
```
Ha! At this point we should brace for questions on why Embark introduces another alias then :)

I'll adjust this paragraph accordingly!
